awx/main/analytics/collectors.py

```diff
@@ -282,33 +283,48 @@ def _copy_table(table, query, path):
+    # `since`, as that would potentially cause gaps in the output
+    # under circumstances where continuous output is desired.
+    from awx.main.signals import disable_activity_stream
+
+    truncate = now() - timedelta(weeks=4)
+    truncate_pk = models.JobEvent.objects.filter(created__lt=truncate).aggregate(Max('pk'))['pk__max']
+    previous_pk = Setting.objects.filter(key='AUTOMATION_ANALYTICS_LAST_EVENT').first()  # Bypass cache
+    previous_pk = max(previous_pk or 0, truncate_pk or 0)
+    final_pk = models.JobEvent.objects.filter(id__gt=previous_pk, created_lte=until).aggregate(Max('pk'))['pk__max']
+
+    events_query = '''COPY (SELECT main_jobevent.id,
                               main_jobevent.created,
                               main_jobevent.modified,
                               main_jobevent.uuid,
                               main_jobevent.parent_uuid,
-                              main_jobevent.event, 
+                              main_jobevent.event,
                               main_jobevent.event_data::json->'task_action' AS task_action,
                               (CASE WHEN event = 'playbook_on_stats' THEN event_data END) as playbook_on_stats,
-                              main_jobevent.failed, 
-                              main_jobevent.changed, 
-                              main_jobevent.playbook, 
+                              main_jobevent.failed,
+                              main_jobevent.changed,
+                              main_jobevent.playbook,
                               main_jobevent.play,
                               main_jobevent.task,
-                              main_jobevent.role, 
-                              main_jobevent.job_id, 
-                              main_jobevent.host_id, 
-                              main_jobevent.host_name
-                              , CAST(main_jobevent.event_data::json->>'start' AS TIMESTAMP WITH TIME ZONE) AS start,
+                              main_jobevent.role,
+                              main_jobevent.job_id,
+                              main_jobevent.host_id,
+                              main_jobevent.host_name,
+                              CAST(main_jobevent.event_data::json->>'start' AS TIMESTAMP WITH TIME ZONE) AS start,
                               CAST(main_jobevent.event_data::json->>'end' AS TIMESTAMP WITH TIME ZONE) AS end,
                               main_jobevent.event_data::json->'duration' AS duration,
                               main_jobevent.event_data::json->'res'->'warnings' AS warnings,
                               main_jobevent.event_data::json->'res'->'deprecations' AS deprecations
-                              FROM main_jobevent 
-                              WHERE (main_jobevent.created > '{}' AND main_jobevent.created <= '{}')
+                              FROM main_jobevent
+                              WHERE (main_jobevent.id > {} AND main_jobevent.id <= {})
```

Comment:
cc @Ladas @jladdjr now that the partitioning PR is taking form, we may want to meet in the middle here and decide how we actually want to do this.  The reason being that in @jladdjr's PR, (as @Ladas pointed out here: ) the primary key of this table is actually now compound (`id`, `job_created`):

https://github.com/ansible/awx/pull/9341#discussion_r584604124

Reply:
@ryanpetrello ok, so for backportable version, we will need to include at least `WHERE main_jobevent.job_created > X` , otherwise this would scan all partitions I believe. But finding the right time window will be a hard, to ensure we send 100% of the events.

Maybe it'll be ok doing index scan of all partitions in this case (given we need this to be reliable). Given we order the index as (id, job_created), PG should use the index if we query just by id (it will just do it in cycle for every table, should be a parallel scan, if that's allowed in the PG config). (@jladdjr good thing we've reordered the index, since (job_created, id) wouldn't be used in this case :-) )

Reply:
Yea @Ladas I think unfortunately we're going to have to do a multi-partition index scan here; there's not really any other way I can think of to be sure we don't miss events if e.g., you don't run the collector for a week.

Reply:
@Ladas is there any sort of duplication detection on the analytics side of things?  Is it an issue if we end up sending an event that's already been received?

Reply:
@ryanpetrello yeah, duplicates are handled on Analytics side (everything is atomic upsert based on primary keys)

Reply:
So @Ladas given this, would it be better in the `devel` branch to just keep our code as-is and switch to `modified`, and for @jbradberry's PK-based approach to be something we *only* track in back ported versions?